samcli/lib/providers/sam_base_provider.py

```diff
@@ -61,14 +68,28 @@ def _extract_lambda_function_code(resource_properties: Dict, code_property_key:
         str
             Representing the local code path
         """
-
         codeuri = resource_properties.get(code_property_key, SamBaseProvider.DEFAULT_CODEURI)
 
         if isinstance(codeuri, dict):
             return SamBaseProvider.DEFAULT_CODEURI
 
         return cast(str, codeuri)
 
+    @staticmethod
+    def _is_code_uri_s3(codeuri: Optional[Any]) -> bool:
+        return (isinstance(codeuri, dict) and ("S3Bucket" in codeuri or "Bucket" in codeuri)) or (
+            isinstance(codeuri, str) and codeuri.startswith("s3://")
+        )
+
+    @staticmethod
+    def _warn_code_extraction(resource_type: str, resource_name: str, code_property: str) -> None:
```

Comment:
Feels slightly weird to have the logging specifically refactored out. Doesn't seem very meaningful on its own.

Would it make sense to (e.g.) instead have a method whose sole purpose is to check whether a resource uses an S3-based `CodeUri`? Something like that should also allow simplifying the 3 very similar use cases in this PR.

Reply:
I think we should do something like we did in https://github.com/aws/aws-sam-cli/pull/2691 (we had similar discussion before).

Reply:
Yeah so I think the whole flow in these methods feels weird as we're essentially dealing with two very similar abstractions (`AWS::Serverless::Function` and `AWS::Lambda::Function`) with vaguely-defined operations (e.g. methods that deal with both abstractions) but still keep repeating similar logic for both.

Not in this scope so I'm happy with this for now, but ideally I'm thinking we would have a well-defined abstraction to deal with functions, e.g. but wrapping them or performing a translation beforehand so we only deal with a subset of well-defined resource types with well-defined operations, so that we don't _need_ all this `if` and `isinstance` ceremony scattered across the codebase.